### ISSUE TITLE
Unable to integrate Odata versioned apis with swagger for Asp.net web application

### ISSUE THREAD

**Issue by vanikulkarniAtFIS** [Microsoft/aspnet-api-versioning#302]
We are using VersionedODataModelBuilder in Http application ( Asp.net web application) 
Its does not have Owin startup hence we pass GlobalConfiguration.Configuration to VersionedODataModelBuilder constructor
Versiong of OData uri works fine as shown in the SwaggerODataWebApiSample
But while integrating this with swagger gives problems:

we create exploere as 

  var apiExplorer = configuration.AddODataApiExplorer(
                options =>
                {
                    options.GroupNameFormat = "'v'VVV";
                    options.SubstituteApiVersionInUrl = true;
                });
				
				
The exception comes at 
 foreach (var group in apiExplorer.ApiDescriptions)
 here apiExplorer.**_ApiDescriptions_** throws excpetion Invalid Operation exception 
 HttpConfiguration.Services.GetApiExplorer		"This operation is not supported by **_'HostedHttpRouteCollection'_**."
 apiExplorer.ApiDescriptions: This operation is not supported by 'HostedHttpRouteCollection'. 

 The reasons could be : GlobalConfiguration.Configuration.Routes is a collection of HostedHttpRoutes 
 whereas in sample application we create new HttpConfigutation
Routes = {System.Web.Http.HttpRouteCollection}
vs
Routes = {System.Web.Http.WebHost.Routing.HostedHttpRouteCollection} 
 We also have normal ( non-odata) web apis as well 
 
Swagger documentation is configured for them too. 
Do you have any such example where odata versioning integrated with swagger and its not Owin

**Comment by vanikulkarniAtFIS** [comment]
Chris, 
Thanks for your quick reply. 
At the moment I cannot upload the source code as its against security policy, and I don't have sample application. 

There no way to modify route collection as its read only property 
What could be the external factors you think is affecting this? 
We are using 

config.Services.Replace(typeof(IApiExplorer), new IntegrationServicesApiExplorer());
IntegrationServicesApiExplorer is searching for controllers in the IntegrationServices folder.
My OData Controllers and normal web api controllers lie in the same folder 
I don't think this should affect it.

**Comment by vanikulkarniAtFIS** [comment]
[Sample Code.zip](https://github.com/Microsoft/aspnet-api-versioning/files/2033955/Sample.Code.zip)


Find few files from our codebase for reference , let me know if you find any problems with this

**Comment by vanikulkarniAtFIS** [comment]
[Swashbuckle.OData.Sample.zip](https://github.com/Microsoft/aspnet-api-versioning/files/2034097/Swashbuckle.OData.Sample.zip)


Please find attached sample code where the issue is reproduced 
We are facing this issue 

System.NotSupportedException was unhandled by user code
  HResult=-2146233067
  Message=This operation is not supported by 'HostedHttpRouteCollection'.
  Source=System.Web.Http.WebHost
  StackTrace:
       at System.Web.Http.WebHost.Routing.HostedHttpRouteCollection.CopyTo(KeyValuePair`2[] array, Int32 arrayIndex)
       at System.Web.Http.HttpRouteCollectionExtensions.GetRouteName(HttpRouteCollection routes, IHttpRoute route)
       at System.Web.Http.HttpConfigurationExtensions.GetODataRootContainer(HttpConfiguration configuration, IHttpRoute route)
       at Microsoft.Web.Http.Description.ODataApiExplorer.ExploreRouteControllers(IDictionary`2 controllerMappings, IHttpRoute route, ApiVersion apiVersion)
       at Microsoft.Web.Http.Description.VersionedApiExplorer.InitializeApiDescriptions()
       at System.Lazy`1.CreateValue()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
       at System.Lazy`1.get_Value()
       at Microsoft.Web.Http.Description.VersionedApiExplorer.get_ApiDescriptions()
  InnerException:

**Comment by vanikulkarniAtFIS** [comment]
Hi Chris,

I have added sample application on https://github.com/Microsoft/aspnet-api-versioning/issues/302 yesterday.
Please have a look at it.
The exact issue is reproduced in the sample app.

This has become quite urgent for us as we want to roll out client side OData apis for version 1.
Looking forward for your quick reply.


Regards,
Vani.

**Comment by vanikulkarniAtFIS** [comment]
Chris,

Any updates on this? Did you get chance to debug sample application?

The information contained in this message is proprietary and/or confidential. If you are not the intended recipient, please: (i) delete the message and all copies; (ii) do not disclose, distribute or use the message in any manner; and (iii) notify the sender immediately. In addition, please be aware that any message addressed to our domain is subject to archiving and review by persons other than the intended recipient. Thank you.

**Comment by commonsensesoftware** [comment]
Sorry for the delayed response, I've been really overloaded lately. Thanks for your patience.  I tried downloading your sample application, but it seems it's no longer there. I don't think GitHub expires this content. Can you try adding it again?

I noticed you said you using:

`config.Services.Replace(typeof(IApiExplorer), new IntegrationServicesApiExplorer());`

Is `IntegrationServicesApiExplorer` an extension of [ODataApiExplorer](../blob/master/src/Microsoft.AspNet.OData.Versioning.ApiExplorer/Description/ODataApiExplorer.cs)? If it's not, then I expect this to fail. There is little-to-no support OData and the **IApiExplorer** out-of-the-box.  I provide a separate package and library to support that. It's normally registered and used as:

`var apiExplorer = config.AddODataApiExplorer();`

This ultimately calls `config.Services.Replace(typeof(IApiExplorer), new ODataApiExplorer(config, options))`. If you want to extend and provide your own customizations, that is a supported scenario. You'll just need to do the registration manually as you're currently doing. If you find a gap or limitation in customizing the current **ODataApiExplorer** implementation, I'm happy to discuss the changes needed to unblock you.

**Comment by vanikulkarniAtFIS** [comment]
Thanks for your reply. In tha sample application we are not replacing IApiExplorer. Attaching the zip file again. Let me know if you can see this 
[Swashbuckle.OData.Sample.zip](https://github.com/Microsoft/aspnet-api-versioning/files/2078908/Swashbuckle.OData.Sample.zip)

**Comment by commonsensesoftware** [comment]
I think I see the problem. You're replacing the **IHttpControllerSelector**, but I don't know why. That's likely going to break API versioning, the API Explorer, or both. It appears that you started trying to build your own API versioning library and then switched.

The **IHttpControllerSelector** is one of the core services that API versioning changes. You can see the implementation [here](../blob/master/src/Microsoft.AspNet.WebApi.Versioning/System.Web.Http/HttpConfigurationExtensions.cs#L46). You can certainly extend and/or customize things, but if you go down that path, you will **not** want to use `AddApiVersioning`. Instead, you'll want to register the services manually. From a routing perspective, there is no difference to API versioning between OData and non-OData routes; that's why they don't have different configuration methods. The things that change for OData is versioning the EDM and adding _version-aware_ OData route conventions.

I've provided a number of OData samples - with and without Swagger. I think you're facing issues from customizations that may no longer be needed (on your side). If there is some customization or configuration you need that isn't supported, I'm happy to help you sort that out. If there's gap in design, then I want to hear about that too.

Thanks

**Comment by commonsensesoftware** [comment]
Yureka! I see what the problem is. You are trying to follow and use code from that [Swashbuckle + OData](https://github.com/rbeauchamp/Swashbuckle.OData). I just noticed something that caught my eye and made me double-check.

**That project is completely different, unrelated, and not compatible with the API versioning libraries.**

This is the first time I've seen this confusion, but I should probably mention something about that on the wiki. This project provides an implementation for **IApiExplorer**. It actually has no dependencies on Swashbuckle or Swagger in any way. Swashbuckle uses the **IApiExplorer** for the Web API implementation to support Swagger.

For an end-to-end working solution with API versioning, OData, and Swagger (a la Swashbuckle), please review the [sample](../tree/master/samples/webapi/SwaggerODataWebApiSample) I provide. You should be able fork it and get it running for your project in no time.

I hope that helps. I'm happy to answer any other questions.

**Comment by commonsensesoftware** [comment]
I was able to repro the problem. Honestly, this is a bug and flaw in the way that the Web API support over IIS was written, but it's unlikely to get fixed any time soon - if ever. The issue is that **RouteCollection** in System.Web.dll doesn't expose the names of routes that it registers. The **HostedHttpRouteCollection** that maps over **RouteCollection** _could_ have mapped these keys during add/remove to the underlying **RouteCollection**, but didn't. Instead, it just throw **NotSupportedException**. These types of runtime integration failures are difficult to detect and/or account for.

I have the fix for this, but it involves very nasty Reflection under the hood. It looks like this scenario is only hit for OData with Web API and hosted on IIS. As much as I wish I could test every permutation of hosting combinations, I'm only one person. Regardless, the fix is relatively small. I should have something out momentarily.

**Comment by commonsensesoftware** [comment]
Version 1.2.2 of the package with the fix has been published. It may take a few minutes to get through the full NuGet vetting process. Feel free to re-open this, if you continue to have an issue. Thanks.

**Comment by vanikulkarniAtFIS** [comment]
Thanks a ton Chris for your efforts and making this so quick. you have saved us from implementing workaround for versioning on OData apis.

**Comment by vanikulkarniAtFIS** [comment]
We are facing some issue. The properties of json are in camel case whereas it should have been the way they are defined in the entity.
By default the name of api controller also appears in camel case 
The sample application (SwaggerOdataWebApiSample) create by you also has this issue.
Any pointers how to resolve this? 
Entity :
 public class Assessment
    {
        public int Id { get; set; }        
        public string AssessmentName { get; set; }
        public long? AssessmentIdentifier { get; set; }
}

We did following setting for Default PropertyNames Convension still it displays Entity Propertis in Camel Case in API responce.
//httpConfiguration.Formatters.JsonFormatter.SerializerSettings.ContractResolver = new DefaultContractResolver();
API Response structure:
{
        Id :123
        assessmentName:"Test"
        assessmentIdentifier:"12345"
}

Issue No 2: 

**Package Required : Swashbuckle.OData**

We are using Swashbuckle.OData package to display Odata Query option on Swagger document. Following configuration is done.

   configuration.EnableSwagger(SwaggerDocsRoute, swagger =>
                    {
					 swagger.CustomProvider(defaultProvider => new ODataSwaggerProvider(defaultProvider, swagger,configuration).Configure(odataConfig =>
                        {
                           odataConfig.IncludeNavigationProperties();
                        }));                        
                    }).EnableSwaggerUi(SwaggerUiRoute, swagger => swagger.EnableDiscoveryUrlSelector());
					
					
After Odata Versiong ODataSwaggerProvider throws Exception - 					
Additional information: Value cannot be null.
{"Value cannot be null.\r\nParameter name: routeData"}
   at System.Net.Http.HttpRequestMessageExtensions.SetRouteData(HttpRequestMessage request, IHttpRouteData routeData)
   at Swashbuckle.OData.Descriptions.SwaggerRouteStrategy.CreateHttpRequestMessage(HttpMethod httpMethod, Operation potentialOperation, SwaggerRoute potentialSwaggerRoute, HttpConfiguration httpConfig) in C:\Users\rbeauchamp\Documents\GitHub\Swashbuckle.OData\Swashbuckle.OData\Descriptions\SwaggerRouteStrategy.cs:line 105
   at Swashbuckle.OData.Descriptions.SwaggerRouteStrategy.GetActionDescriptors(HttpMethod httpMethod, Operation potentialOperation, SwaggerRoute potentialSwaggerRoute, HttpConfiguration httpConfig) in C:\Users\rbeauchamp\Documents\GitHub\Swashbuckle.OData\Swashbuckle.OData\Descriptions\SwaggerRouteStrategy.cs:line 67
   at Swashbuckle.OData.Descriptions.SwaggerRouteStrategy.GetActionDescriptors(SwaggerRoute potentialSwaggerRoute, HttpConfiguration httpConfig) in C:\Users\rbeauchamp\Documents\GitHub\Swashbuckle.OData\Swashbuckle.OData\Descriptions\SwaggerRouteStrategy.cs:line 52
   at Swashbuckle.OData.Descriptions.SwaggerRouteStrategy.<>c__DisplayClass3_0.<Generate>b__1(SwaggerRoute potentialSwaggerRoute) in C:\Users\rbeauchamp\Documents\GitHub\Swashbuckle.OData\Swashbuckle.OData\Descriptions\SwaggerRouteStrategy.cs:line 41
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<DistinctIterator>d__63`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Swashbuckle.OData.CollectionExtentions.ToCollection[T](IEnumerable`1 source) in C:\Users\rbeauchamp\Documents\GitHub\Swashbuckle.OData\Swashbuckle.OData\CollectionExtentions.cs:line 77
   at Swashbuckle.OData.Descriptions.ODataApiExplorer.GetApiDescriptions() in C:\Users\rbeauchamp\Documents\GitHub\Swashbuckle.OData\Swashbuckle.OData\Descriptions\ODataApiExplorer.cs:line 43
   at System.Lazy`1.CreateValue()
      
   Issue No -3   
   Compile time error is occured for Odata ApiExplorer.   
   {"Object reference not set to an instance of an object."}
      at Microsoft.Web.Http.Description.TypeExtensions.<>c__DisplayClass10_0.<GetStructuredType>b__0(IEdmStructuredType t)
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.Web.Http.Description.TypeExtensions.GetStructuredType(Type type, IEdmModel model, IAssembliesResolver assembliesResolver)
   at Microsoft.Web.Http.Description.TypeExtensions.SubstituteIfNecessary(Type type, IServiceProvider serviceProvider, IAssembliesResolver assembliesResolver, ModelTypeBuilder modelTypeBuilder)
   at Microsoft.Web.Http.Description.ODataApiExplorer.CreateParameterDescriptionFromBinding(HttpParameterBinding parameterBinding, IServiceProvider serviceProvider, IAssembliesResolver assembliesResolver)
   at Microsoft.Web.Http.Description.ODataApiExplorer.CreateParameterDescriptions(HttpActionDescriptor actionDescriptor, IHttpRoute route)
   at Microsoft.Web.Http.Description.ODataApiExplorer.ExploreRouteActions(IHttpRoute route, HttpControllerDescriptor controllerDescriptor, IHttpActionSelector actionSelector, Collection`1 apiDescriptions, ApiVersion apiVersion)
   at Microsoft.Web.Http.Description.ODataApiExplorer.ExploreRouteControllers(IDictionary`2 controllerMappings, IHttpRoute route, ApiVersion apiVersion)
   at Microsoft.Web.Http.Description.VersionedApiExplorer.InitializeApiDescriptions()
   at System.Lazy`1.CreateValue()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Lazy`1.get_Value()
   at Microsoft.Web.Http.Description.VersionedApiExplorer.get_ApiDescriptions()

**Comment by vanikulkarniAtFIS** [comment]
There are 3 new issues we found, let me know if I need to create new issue

**Comment by commonsensesoftware** [comment]
### Issue 1
OData doesn't use the default JSON **MediaTypeFormatter**; it uses its own. Camelcase is very common for JSON. I suspect you are using the configuration in the examples I provide, which enable this. Simply remove this configuration. For example:

```c#
var modelBuilder = new VersionedODataModelBuilder( configuration )
{
    // configuration shown in example using camelcase
    // ModelBuilderFactory = () => new ODataConventionModelBuilder().EnableLowerCamelCase(),

    // default configuration using source case, which usually Pascalcase
    ModelBuilderFactory = () => new ODataConventionModelBuilder(),

    // remaining configuration ...
};
```

### Issue 2
I really know very little about the **Swashbuckle.OData** library or project. I don't use it or support it. You _should_ have everything you need with the OData API versioning, OData API Explorer, and vanilla Swashbuckle. If there is something not supported that you need, I'd be interested in hearing about it to determine if a new feature is required. My recommendation would be to remove **Swashbuckle.OData** all together. It shouldn't be needed.

### Issue 3
I'm not sure what this is. Does it repro in the example that you previously provided? If you can provide a copy of the most version of that sample where the scenario repros, I'll take a look. It's possible there's a bug.
